LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js

```diff
@@ -151,6 +151,126 @@ async function dispatchTouchActions(actions, options = { insertPauseAfterPointer
+    maxDuration = 0;
+    for (let action of actions) {
+        let duration;
+        if (action.type === "pause" || action.type === sourceToActionWithDurationMap[sourceType])
+            duration = action.duration;
+        if (duration !== undefined) {
+            if (duration > maxDuration)
+                maxDuration = duration;
+        }
+    }
+    return maxDuration;
+}
+
+async function renderingUpdate()
+{
+    await new Promise(requestAnimationFrame);
+    await new Promise(resolve => setTimeout(resolve, 0));
+}
+
+async function waitForScrollCompletion()
+{
+    if (testRunner.isIOSFamily) {
+        await new Promise(resolve => {
+            testRunner.runUIScript(`
+                (function() {
+                    uiController.didEndScrollingCallback = function() {
+                        uiController.uiScriptComplete();
+                    }
+                })()`, resolve);
+        });
+        return renderingUpdate();
+    }
+
+    return new Promise(resolve => {
+        eventSender.callAfterScrollingCompletes(() => {
+            requestAnimationFrame(resolve);
+        });
+    });
+}
+
+async function ensurePresentationUpdate()
+{
+    if (!testRunner.isWebKit2)
+        return renderingUpdate();
+
+    return new Promise(resolve => {
+        testRunner.runUIScript(`
+            uiController.doAfterPresentationUpdate(function() {
+                uiController.uiScriptComplete();
+            });`, resolve);
+    });
+}
+
+/**
+ *
+ * @param {object[]} actions
+ * @returns Promise<undefined>
+ */
+async function dispatchWheelActions(actions)
+{
+    if (!window.eventSender)
+        throw new Error("window.eventSender is undefined.");
+
+    for (let action of actions) {
+        switch (action.type) {
+        case "pause":
+            logDebug(`pause(${action.duration})`);
+            await pause(action.duration);
+            break;
+        case "scroll":
+            // FIXME(261810): Follow tick-based event dispatch logic rather than sending scroll events instantaneously
+            // https://w3c.github.io/webdriver/#dfn-perform-a-scroll
+            let { x, y, origin, deltaX, deltaY, duration } = action;
+            if ((x < 0) || (x > window.innerWidth) || (y < 0) || (y > window.innerHeight))
+                throw new Error('Move target out of bounds');
+            if (duration === undefined)
+                duration = computeTickDuration(actions, "wheel");
+            if (origin instanceof Element) {
+                const bounds = origin.getBoundingClientRect();
+                logDebug(() => `${origin.id} [${bounds.left}, ${bounds.top}, ${bounds.width}, ${bounds.height}]`);
+                x += bounds.left + 1;
+                y += bounds.top + 1;
+            }
+            const scrollEvents = [
+                {
+                    type : "wheel",
+                    viewX : x,
+                    viewY : y,
+                    deltaX : -deltaX,
```

Comment:
Can you explain the negation? Are you sure X should be negated?

Reply:
I will double check this. I assumed both deltas need negation because event streams elsewhere in LayoutTests negated `deltaX` too, but maybe I misunderstood the point. For example in [`fast/scrolling/mac/overflow-hidden-on-one-axis-async-overflow.html`](https://github.com/WebKit/WebKit/blob/50cfb6afc83f9e6eb1b80c4a217b417648176145/LayoutTests/fast/scrolling/mac/overflow-hidden-on-one-axis-async-overflow.html#L65-L74)

```js
                type : "wheel",
                viewX : 100,
                viewY : 100,
                deltaX : -10,
                phase : "began"
            },
            {
                type : "wheel",
                deltaX : -100,
                phase : "changed"
```

Reply:
Yes, the `deltaX` value should be negated too. This is what the [WebDriver spec prescribes about performing a scroll](https://www.w3.org/TR/webdriver2/#dfn-perform-implementation-specific-action-dispatch-steps:~:text=Perform%20implementation%2Dspecific%20action%20dispatch%20steps%20on%20browsing%20context%20equivalent%20to%20wheel%20scroll%20at%20viewport%20x%20coordinate%20x%2C%20viewport%20y%20coordinate%20y%2C%20deltaX%20value%20delta%20x%2C%20deltaY%20value%20delta%20y%2C%20in%20accordance%20with%20the%20requirements%20of%20%5BUI%2DEVENTS%5D.) (emphasis mine):

> Perform implementation-specific action dispatch steps on browsing context equivalent to wheel scroll at viewport x coordinate x, viewport y coordinate y, deltaX value delta x, deltaY value delta y, _in accordance with the requirements of [UI-EVENTS]_.

The UI Events spec says this [about the delta values for wheel events](https://www.w3.org/TR/uievents/#events-wheelevents:~:text=the%20sign%20of%20the%20delta%20SHOULD%20be%20given%20by%20a%20right%2Dhand%20coordinate%20system%20where%20positive%20X%2C%20Y%2C%20and%20Z%20axes%20are%20directed%20towards%20the%20right%2Dmost%20edge%2C%20bottom%2Dmost%20edge%2C%20and%20farthest%20depth%20(away%20from%20the%20user)%20of%20the%20document%2C%20respectively.):

> ... the sign of the delta SHOULD be given by a right-hand coordinate system where positive X, Y, and Z axes are directed towards the right-most edge, bottom-most edge, and farthest depth (away from the user) of the document, respectively.

---

According to my investigation, a wheel event's negative `deltaX` value (in the event stream) reflects a scroll towards the right-most edge of the document in WKTR, whereas the `deltaX` argument to the WPT actions interface represents the opposite (positive tends to the right-most edge). This is why we should negate both `deltaX` and `deltaY`.